final_project/flaskr/schema.sql

```diff
@@ -2,50 +2,39 @@
 /* Why is 'status' blue 
 There are actually 3 states a bikerack can be in: "approved", "pending", and
 "rejected".
-for each vote also include it's time/date that it was voted
-separate table for votes, foreign key on bike rack id
-and this separate table has the data for voting: time/date, type of vote ("upvote" or "downvote"), index
-on the bikerack id (or time/date)
+separate table for votes, has foreign key on bike rack id
+and this separate table has the data for voting: type of vote (1 ("upvote") or -1 ("downvote")),
+user_id, index on the bikerack id (or user_id)
 The relationship between bikeracks and votes is one to many, one bikerack
 to many votes
 The bikeracks table is the parent table, and is the table that the
 foreign key constraint refers to.
 The votes table is the child table, which is the table that a foreign key
 consraint is applied.
 The rack_id in bikeracks is the parent key, the rack_id column in
-votes is called the child key.
-In sqlite the datatype of a value is associated with the value itself,
-not with its container */
-
-/* TODO Should I have an upvote count and downvote count in the bikeracks database?
-Then you don't have to count up the number of upvotes for a particular rack in the 
-votes table every time. */
+votes is called the child key.*/
 
 DROP TABLE IF EXISTS bikeracks;
 DROP TABLE IF EXISTS votes;
-DROP TABLE IF EXISTS users;
 DROP TABLE IF EXISTS bikeracks_history;
 DROP TABLE IF EXISTS votes_history;
-DROP TABLE IF EXISTS users_history; 
 
 CREATE TABLE bikeracks (
     rack_id INTEGER PRIMARY KEY AUTOINCREMENT,
     status TEXT DEFAULT "pending",
     latitude REAL NOT NULL,
     longitude REAL NOT NULL,
     address TEXT,
+    upvote_count INTEGER NOT NULL DEFAULT 0,
+    downvote_count INTEGER NOT NULL DEFAULT 0,
     CONSTRAINT coordinates UNIQUE(latitude, longitude),
     CHECK (status in ("approved", "pending", "rejected"))
     
 );
 
-/* vote_timestamp is type INTEGER as unix time, number of seconds since 1970-01-01 00:00:00 UTC
-TODO A users table is going to change things here. a vote will be tied to a user
-potential users schema: id, username, password, email, password salt?, password hashed?, common name, date created
-and the users id will be a foreign key on 
-the votes table because:
+/* 
 so when a user upvotes a bikerack, first we check, have they already upvoted this
-particular bikerack? and we check this by looking up all votes using their user id
+particular bikerack? and we check this by looking up all votes using their user_id
 and checking if the rack_id is already there for this rack. Because the front end side
 doesn't know anything about rack ids, we'd first need to look up this rack in the
```

Comment:
This works as long as the roundtrip through javascript doesn't change the float value at all (float equality is hard to get right). You might consider limiting the lat and long to N decimal places to account for any lossy floating point operations that might happen when converting from `REAL` to python `float` to javascript `Number` to python `float` to database `REAL` again.

Alternatively, _could_ the frontend know about the rack_id? Why not? It has to load the racks to display them before anyone can vote on them anyway.

Reply:
I think it would be easier if the front end knows about the rack_id.